### ISSUE TITLE
NAPTR Regex parse wrong double backslash according to RFC 2915

### ISSUE THREAD

**Issue by TeoV** [miekg/dns#974]
Greetings,

Today we try some scenario and found that the double backslashed are wrongly parsed.
According to RFC 2915  if you want to have a backslash you must escape that backslash with another backslash. 
So in theory if the server send this string as a NAPTR Regex  `/urn:cid:.+@([^\\.]+\\.)(.*)$/\\2/i` the client should receive `/urn:cid:.+@([^\.]+\.)(.*)$/\2/i` ( (Example1: https://tools.ietf.org/html/rfc2915#section-7.1 )
In our test we send the exact same string as regex : 
`/urn:cid:.+@([^\\.]+\\.)(.*)$/\\2/i`
But the client receive the following : 
`/urn:cid:.+@([^\\.]+\\.)(.*)$/\\2/i`
In case that  server send the request with 1 backslash the client receive the request without backslash.
What server send : 
`/urn:cid:.+@([^\.]+\.)(.*)$/\2/i`
What client receive : 
`/urn:cid:.+@([^.]+.)(.*)$/2/i`

Additional information, in our code, we use a dns.ResponseWriter to send the message to the client ( https://github.com/cgrates/cgrates/blob/master/agents/libdns.go#L110-L117 ) 

Thanks,
TeoV

**Comment by tmthrgd** [comment]
What client are you using? Are you sure the data on the wire is actually wrong? It could just be the client escaping it before displaying it.

**Comment by TeoV** [comment]
Hi @tmthrgd,

In our test, we use the client from this library 
https://github.com/cgrates/cgrates/blob/master/agents/dnsagent_it_test.go#L107-L116
Basically, the client sends a message to the server, server process and compose the new message and send back to the client. After that, we read the message from the server and check if the data are the one that we expected.

Thanks,
TeoV

**Comment by miekg** [comment]
It this because it's a Go strings, to you need to re-add the escapes? i.e. `fmt.Printf("%q", ...)` ?

**Comment by TeoV** [comment]
Hi @miekg,

On GitHub we have %q but I tested on my local machine with %+v ( the result that I posted are with %+v ).

Thanks,
TeoV

**Comment by TeoV** [comment]
If it's ok tomorrow I will do a small test to reproduce the problem and I will post it here so you can troubleshoot it easily?

**Comment by TeoV** [comment]
Hello,

I attach a basic server client to reproduce the problem.
[BasicServerClientExample.zip](https://github.com/miekg/dns/files/3206097/BasicServerClientExample.zip)
If you need any information don't hesitate to ask.

Thanks,
TeoV

**Comment by miekg** [comment]
~~~
--- server.go.orig	2019-05-23 20:52:09.848063790 +0100
+++ server.go	2019-05-23 20:52:17.363975723 +0100
@@ -27,7 +27,7 @@
 	}
 	//we tried the following string :
 	//1)
-	itm := "/urn:cid:.+@([^\\.]+\\.)(.*)$/\\2/i"
+	itm := `/urn:cid:.+@([^\\.]+\\.)(.*)$/\\2/i`
 	// in this case the client will receive
 	// 		/urn:cid:.+@([^.]+.)(.*)$/2/i
 	//2)
~~~
seem to fix it

**Comment by TeoV** [comment]
Hi @miekg,

We tried to change the string as you suggested but the client receives a wrong format.
We found that the problem comes from Pack and Unpack methods.
Please run this test and you will notice the difference.

```
func TestErrorPackUnpack(t *testing.T) {
	m := new(Msg)
	m.SetQuestion("3.6.9.4.7.1.7.1.5.6.8.9.4.e164.arpa.", TypeNAPTR)
	rply := new(Msg)
	rply.SetReply(m)
	//add answer as NAPTR
	if len(rply.Answer) == 0 {
		rply.Answer = append(rply.Answer,
			&NAPTR{
				Hdr: RR_Header{
					Name:   m.Question[0].Name,
					Rrtype: m.Question[0].Qtype,
					Class:  ClassINET,
					Ttl:    60},
			},
		)
	}
	itm := `/urn:cid:.+@([^\\.]+\\.)(.*)$/\\2/i`
	rply.Answer[len(rply.Answer)-1].(*NAPTR).Regexp = itm
	// Pack msg and then unpack into msg2
	buf, err := rply.Pack()
	if err != nil {
		t.Fatalf("rply.Pack failed: %v", err)
	}
	var msg2 Msg
	if err := msg2.Unpack(buf); err != nil {
		t.Fatalf("msg2.Unpack failed: %v", err)
	}
	if msg2.Answer[0].(*NAPTR).Regexp != `/urn:cid:.+@([^\.]+\.)(.*)$/\2/i` {
		t.Errorf("\nExpecting : <%+v>, but received <%+v> \n", `/urn:cid:.+@([^\.]+\.)(.*)$/\2/i`, msg2.Answer[0].(*NAPTR).Regexp)
	}
}
```

Thanks,
TeoV

**Comment by TeoV** [comment]
Hi @miekg,

Any chance that you have a look at this test? 

Thanks,
TeoV

**Comment by tmthrgd** [comment]
The data you’re putting in is different to what your test is expecting out, see:
```
`/urn:cid:.+@([^\\.]+\\.)(.*)$/\\2/i`
`/urn:cid:.+@([^\.]+\.)(.*)$/\2/i`
```

I haven’t run the test, but that doesn’t seem to be an issue with this package.

**Comment by TeoV** [comment]
Hi @tmthrgd,

Please have a look here at the RFC ( https://tools.ietf.org/html/rfc2915#section-7.1 ) 
It says and I quote : 

> ` Since '\' is the escape character, literal occurances of a backslash must be escaped by another backslash.  For the case of the cid.urn.arpa record above, the regular expression entered into the master file should be  "/urn:cid:.+@([^\\.]+\\.)(.*)$/\\2/i".  When the client code actually   receives the record, the pattern will have been converted to "/urn:cid:.+@([^\.]+\.)(.*)$/\2/i".`

So, in theory, the server " pack " a message that contains double backslashes  ( https://github.com/miekg/dns/blob/master/msg.go#L504-L539 ) send to the client and the client " unpack" the message and expect to have one backslash  ( https://github.com/miekg/dns/blob/master/msg_helpers.go#L263-L286). 

Thanks,
TeoV

**Comment by tmthrgd** [comment]
Okay I see where you’re getting confused, that section of the RFC specifically mentions “the master file” which we call the zone file—that is parsed by the `ZoneParser` type or `NewRR`. Parsing zone files *does* unescape one level of backslashes and then it produces a `dns.RR` that’s been unescaped.

Pack -> unpack needs to roundtrip so the data must end up the same as it started. That’s not what the RFC is talking about there.

As far as I can tell, there is nothing actionable or wrong with this package. The unescaping the RFC is talking about happens in the `ZoneParser`, but if you’re constructing `NAPTR` types directly, you need to provide the data as you want the client to receive it.

**Comment by TeoV** [comment]
Hi @miek, @tmthrgd,

I think I express myself wrong: 
We try to send something and we expect to receive the same thing : 
To notice this difference please run this test : 

```
func TestErrorPackUnpack(t *testing.T) {
	m := new(Msg)
	m.SetQuestion("3.6.9.4.7.1.7.1.5.6.8.9.4.e164.arpa.", TypeNAPTR)
	rply := new(Msg)
	rply.SetReply(m)
	//add answer as NAPTR
	if len(rply.Answer) == 0 {
		rply.Answer = append(rply.Answer,
			&NAPTR{
				Hdr: RR_Header{
					Name:   m.Question[0].Name,
					Rrtype: m.Question[0].Qtype,
					Class:  ClassINET,
					Ttl:    60},
			},
		)
	}
	itm := `/urn:cid:.+@([^\.]+\.)(.*)$/\2/i`
	rply.Answer[len(rply.Answer)-1].(*NAPTR).Regexp = itm
	// Pack msg and then unpack into msg2
	buf, err := rply.Pack()
	if err != nil {
		t.Fatalf("rply.Pack failed: %v", err)
	}
	var msg2 Msg
	if err := msg2.Unpack(buf); err != nil {
		t.Fatalf("msg2.Unpack failed: %v", err)
	}
	if msg2.Answer[0].(*NAPTR).Regexp != itm {
		t.Errorf("\nExpecting : <%+v>, but received <%+v> \n", itm, msg2.Answer[0].(*NAPTR).Regexp)
	}
}
```

Thanks,
TeoV

**Comment by miekg** [comment]
thanks. That test is useful.

The issue as that in `packTxtString` we check for `\DDD` sequences, but this eats the `\` in this case. And that is indeed a bug.

**Comment by tmthrgd** [comment]
@miekg Is that a bug though? By the looks of it, the field is expecting to take an escaped form with the \ character as the escape character. In that case it should be eating the \ character.

Because the code needs to roundtrip, whatever unpack does, pack needs to do the opposite.

@TeoV Is there some public server (*not* based on miekg/dns or CoreDNS—perhaps BIND) that sends these records to test against?

Basically whatever the unpack code produces when pointed at a different server is what the pack code needs to consume.

What is the result of pointing dig at a miekg/dns server? And how does it compare to pointing dig at a different server?

I think this is actually working as intended and manual unescaping/escaping needs to be done by the consumer. Although I could be very wrong here.

**Comment by miekg** [comment]
From my reading of 1035 character-string should not touch it (except maybe
for \DDD). We re-add the backslashes in unpackString. Probably should do
neither and I doubt this is interoperable with other implementations, but
that needs checking.

**Comment by tmthrgd** [comment]
Changing that would be backwards incompatible though. I also think a sometimes escaping, sometimes not escape character is a recipe for disaster. Either we support escaping or we don’t, and I think that ship has sailed. 

I really do think this is working as intended, even if it may be a tad unfortunate and awkward.

I still want to be able to compare to other servers and dig though.

**Comment by gibson042** [comment]
Backslashes in presentation-format character-strings _should_ be consumed, because they are escape characters. Per [RFC 1035 section 5.1](https://tools.ietf.org/html/rfc1035#page-35), they must be followed by either three decimal digits `000` through `255` (representing any single octet) or by a non-digit character (representing the value of that literal character without special meaning). The wire-format data of a character-string can only contain a backslash where the corresponding presentation format (e.g., from a zone file) has escaped it like either `\\` or `\092`.

For the example, please read https://tools.ietf.org/html/rfc3404#page-11 carefully and note that NAPTR presentation-format RDATA should look like this:
```
(
  ; ORDER
  100

  ; PREFERENCE
  10

  ; FLAGS
  ""

  ; SERVICES (parameters)
  ""

  ; REGEXP
  ; sed-like <delim-char><POSIX ERE><delim-char><replacement><delim-char><flags>
  ; cf. https://tools.ietf.org/html/rfc3402#section-3.2
  "!^cid:.+@([^\\.]+\\.)(.*)$!\\2!i"

  ; REPLACEMENT (domain)
  .
)
```
(or equivalently `100 10 "" "" "!^cid:.+@([^\092.]+\092.)(.*)$!\0922!i" .`)

to yield wire-format RDATA (cf. https://tools.ietf.org/html/rfc3403#section-4.1 ) like this:

octets | field | type | value
-- | -- | -- | --
00 64 | ORDER | uint16 | `100`
00 0A | PREFERENCE | uint16 | `10`
00 | FLAGS | \<character-string> | 
00 | SERVICES | \<character-string> | 
1D 21 5E 63 69 64 3A 2E 2B 40 28 5B 5E 5C 2E 5D 2B 5C 2E 29 28 2E 2A 29 24 21 5C 32 21 69 | REGEXP | \<character-string> | `!^cid:.+@([^\.]+\.)(.*)$!\2!i`
00 | REPLACEMENT | \<domain-name> | `.`

**Comment by TeoV** [comment]
Hi @miekg,

By any chance have you managed to take a look at this?

Thanks,
TeoV

**Comment by miekg** [comment]
No I haven't. Looks WAI to me
